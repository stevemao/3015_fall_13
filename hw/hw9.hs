module Boolean where 

  type Boolean = Either () ()

  true, false :: Boolean
  true = Left ()
  false = Right ()
 
  ifthenelse :: Boolean -> a -> a  -> a
  ifthenelse b e1 e2 = either (\() -> e1) (\() -> e2) b


--   (/\),(\/),(.=>) :: Boolean -> Boolean -> Boolean
