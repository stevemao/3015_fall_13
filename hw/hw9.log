This is pdfTeX, Version 3.14159-14h-released-20010417 (Web2C 7.3.3.1) (format=pdflatex 2006.5.13)  26 SEP 2006 16:47
**hw9
(./hw9.tex{/usr/share/texmf/pdftex/config/pdftex.cfg}
LaTeX2e <2001/06/01>
Babel <v3.7h> and hyphenation patterns for american, french, german, ngerman, i
talian, nohyphenation, loaded.
(/usr/share/texmf/tex/latex/base/article.cls
Document Class: article 2001/04/21 v1.4e Standard LaTeX document class
(/usr/share/texmf/tex/latex/base/size11.clo
File: size11.clo 2001/04/21 v1.4e Standard LaTeX file (size option)
)
\c@part=\count79
\c@section=\count80
\c@subsection=\count81
\c@subsubsection=\count82
\c@paragraph=\count83
\c@subparagraph=\count84
\c@figure=\count85
\c@table=\count86
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)
\@input{header.aux}
\openout2 = `header.aux'.

(./header.tex (/usr/share/texmf/tex/latex/amsfonts/amssymb.sty
Package: amssymb 2002/01/22 v2.2d

(/usr/share/texmf/tex/latex/amsfonts/amsfonts.sty
Package: amsfonts 2001/10/25 v2.2f
\@emptytoks=\toks14
\symAMSa=\mathgroup4
\symAMSb=\mathgroup5
LaTeX Font Info:    Overwriting math alphabet `\mathfrak' in version `bold'
(Font)                  U/euf/m/n --> U/euf/b/n on input line 132.
)) (./remark.sty)
(/usr/share/texmf/tex/latex/html/html.sty
(/usr/share/texmf/tex/latex/hyperref/hyperref.sty
Package: hyperref 2002/08/04 v6.72v Hypertext links for LaTeX

(/usr/share/texmf/tex/latex/graphics/keyval.sty
Package: keyval 1999/03/16 v1.13 key=value parser (DPC)
\KV@toks@=\toks15
)
\@linkdim=\dimen103
\Hy@linkcounter=\count87
\Hy@pagecounter=\count88

(/usr/share/texmf/tex/latex/hyperref/pd1enc.def
File: pd1enc.def 2002/08/04 v6.72v Hyperref: PDFDocEncoding definition (HO)
)
Package hyperref Info: Bookmarks ON on input line 1797.
Package hyperref Info: Hyper figures OFF on input line 1816.
Package hyperref Info: Link nesting OFF on input line 1821.
Package hyperref Info: Hyper index ON on input line 1824.
Package hyperref Info: Plain pages ON on input line 1829.
Package hyperref Info: Backreferencing OFF on input line 1836.

Implicit mode ON; LaTeX internals redefined
(/usr/share/texmf/tex/latex/html/url.sty
Package: url 1999/03/02  ver 1.4  Verb mode for urls, email addresses, and file
 names
)
LaTeX Info: Redefining \url on input line 2048.
\Fld@menulength=\count89
\Field@Width=\dimen104
\Fld@charsize=\dimen105
\Choice@toks=\toks16
\Field@toks=\toks17
Package hyperref Info: Hyper figures OFF on input line 2506.
Package hyperref Info: Link nesting OFF on input line 2511.
Package hyperref Info: Hyper index ON on input line 2514.
Package hyperref Info: backreferencing OFF on input line 2521.
Package hyperref Info: Link coloring OFF on input line 2526.
\c@Item=\count90
\c@Hfootnote=\count91
)
*hyperref using default driver hpdftex*
(/usr/share/texmf/tex/latex/hyperref/hpdftex.def
File: hpdftex.def 2002/08/04 v6.72v Hyperref driver for pdfTeX

(/usr/share/texmf/tex/latex/psnfss/pifont.sty
Package: pifont 2001/06/04 PSNFSS-v8.2 Pi font support (SPQR) 
LaTeX Font Info:    Try loading font information for U+pzd on input line 62.

(/usr/share/texmf/tex/latex/psnfss/upzd.fd
File: upzd.fd 2001/06/04 font definitions for U/pzd.
)
LaTeX Font Info:    Try loading font information for U+psy on input line 63.

(/usr/share/texmf/tex/latex/psnfss/upsy.fd
File: upsy.fd 2001/06/04 font definitions for U/psy.
))
\Fld@listcount=\count92
\@outlinefile=\write3
)
Package: html 1999/07/19 v1.38 hypertext commands for latex2html (nd, hws, rrm)

\c@lpart=\count93
\c@lchapter=\count94
\c@chapter=\count95
\c@lsection=\count96
\c@lsubsection=\count97
\c@lsubsubsection=\count98
\c@lparagraph=\count99
\c@lsubparagraph=\count100
\c@lsubsubparagraph=\count101
\ptrfile=\write4
)
(/home/faculty/jlc/papers/tex-stuff/bussproofs.sty
Proof Tree (bussproofs) style macros. Version 0.6c.
\theLevel=\count102
\myMaxLevel=\count103
\myBoxA=\box26
\myBoxB=\box27
\myBoxC=\box28
\myBoxD=\box29
\myBoxLL=\box30
\myBoxRL=\box31
\thisAboveSkip=\dimen106
\thisBelowSkip=\dimen107
\newScoreStart=\dimen108
\newScoreEnd=\dimen109
\newCenter=\dimen110
\displace=\dimen111
\leftLowerAmt=\dimen112
\rightLowerAmt=\dimen113
\scoreHeight=\dimen114
\scoreDepth=\dimen115
)
\c@theorem=\count104
\c@conjecture=\count105
\c@corollary=\count106
\c@example=\count107
\c@fact=\count108
\c@lemma=\count109
\c@definition=\count110
\c@claim=\count111
\c@remark=\count112
\c@slogan=\count113
\c@axiom=\count114
\c@note=\count115
\c@problem=\count116
\c@exercise=\count117
) (./hw9.aux)
\openout1 = `hw9.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for PD1/pdf/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
Package hyperref Info: Link coloring OFF on input line 5.
 (/usr/share/texmf/tex/latex/hyperref/nameref.sty
Package: nameref 2001/01/27 v2.19 Cross-referencing by name of section
\c@section@level=\count118
)
LaTeX Info: Redefining \ref on input line 5.
LaTeX Info: Redefining \pageref on input line 5.
 (./hw9.out)
(./hw9.out)
\openout3 = `hw9.out'.


Underfull \hbox (badness 10000) in paragraph at lines 6--7

 []

LaTeX Font Info:    Try loading font information for U+msa on input line 37.
(/usr/share/texmf/tex/latex/amsfonts/umsa.fd
File: umsa.fd 2002/01/19 v2.2g AMS font definitions
)
LaTeX Font Info:    Try loading font information for U+msb on input line 37.

(/usr/share/texmf/tex/latex/amsfonts/umsb.fd
File: umsb.fd 2002/01/19 v2.2g AMS font definitions
) [1



{/usr/share/texmf/dvips/config/pdftex.map}]
LaTeX Font Info:    Try loading font information for OMS+cmtt on input line 73.

LaTeX Font Info:    No file OMScmtt.fd. on input line 73.


LaTeX Font Warning: Font shape `OMS/cmtt/m/n' undefined
(Font)              using `OMS/cmsy/m/n' instead
(Font)              for symbol `textbraceleft' on input line 73.

! pdfTeX warning (ext4): destination with the same identifier (name{exercise.1}
) has been already used, duplicate ignored
<to be read again> 
                   \penalty 
l.97 \exercise
              {Using the definition of {\tt{ifthenelse, true}} and {\tt{fals...

Overfull \hbox (5.50493pt too wide) in paragraph at lines 112--116
/cmr10@10.95pt/You may use the de-struc-tor /cmtt10@10.95pt/either /cmr10@10.95
pt/di-rectly if you like, or de-fine the func-tions in terms of /cmtt10@10.95pt
/ifthenelse/cmr10@10.95pt/.
 []

[2] (./hw9.aux)

LaTeX Font Warning: Some font shapes were not available, defaults substituted.

 ) 
Here is how much of TeX's memory you used:
 2778 strings out of 50453
 34854 string characters out of 690528
 82965 words of memory out of 1500001
 5757 multiletter control sequences out of 10000+25000
 12780 words of font info for 49 fonts, out of 400000 for 1000
 14 hyphenation exceptions out of 10000
 38i,8n,36p,236b,231s stack positions out of 3000i,500n,3000p,50000b,5000s
</usr/share/texmf/fonts/type1/bluesky/cm/cmsy9.pfb></usr/share/texmf/fonts/ty
pe1/bluesky/cm/cmtt9.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmr9.pfb></us
r/share/texmf/fonts/type1/bluesky/cm/cmr6.pfb></usr/share/texmf/fonts/type1/blu
esky/cm/cmr8.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmti10.pfb></usr/shar
e/texmf/fonts/type1/bluesky/cm/cmsy10.pfb></usr/share/texmf/fonts/type1/bluesky
/cm/cmtt10.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmbx12.pfb></usr/share/
texmf/fonts/type1/bluesky/cm/cmr10.pfb></usr/share/texmf/fonts/type1/bluesky/cm
/cmbx10.pfb>
Output written on hw9.pdf (2 pages, 68738 bytes).
