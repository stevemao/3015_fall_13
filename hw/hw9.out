\BOOKMARK [1][-]{section.1}{Disjoint Unions}{}
\BOOKMARK [1][-]{section.2}{The Unit Type}{}
\BOOKMARK [1][-]{section.3}{Another encoding of Booleans using Disjoint Unions}{}
