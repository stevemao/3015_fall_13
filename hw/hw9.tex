\documentclass[11pt]{article}

\include{header}

\begin{document}
\homework{9}{28 September}

\section{Disjoint Unions}

\exercise{Reread section 2.5 of Bird.}

The Hugs implementation of the the type {\tt{Either}} differs from the
older version of Haskell described in the book in that the following
code is in the prelude.

\begin{program*}
\> data  Either a b  =  Left a | Right b   deriving (Eq, Ord, Read, Show) \\
\>  \\
\> either               :: (a -> c) -> (b -> c) -> Either a b -> c \\
\> either f g (Left x)  =  f x \\
\> either f g (Right y) =  g y 
\end{program*}

So, in Hugs, we use the destructor {\tt{either}} rather than {\tt{case}} as described in Bird. In class, we said:

\begin{program*}
\> case                :: ((a -> c), (b -> c)) -> (Either a b) -> c \\
\> case (f,g) (Left x)  =  f x \\
\> case (f,g) (Right y) =  g y 
\end{program*}

So, {\tt{either}} is just like {\tt{case}}, except it is  fully curried.

\section{The Unit Type}

There is a type of tuples of length zero written {\tt{()}}.  There are
two inhabitants of theis type, $\bot$ (which inhabits every type) and
the element {\tt{()}}.  This can be a bit confusing since, for the
unit type, the element of the unit type has the same syntax as the
type.

Now, we can make a constant into a function by making it a function of
the unit type.
\begin{program*}
\> pifun :: () -> Float \\
\> pifun () = 3.14159
\end{program*}
In the signature for {\tt{pifun}}, ``{\tt{()}}'' denotes the unit type.  In the
definition of {\tt{pifun}}, ``{\tt{()}}'' denotes the single element of the
unit type.

\newpage
\section{Another encoding of Booleans using Disjoint Unions}

Now, conside the following module.

\begin{program*}
\> module Boolean where  \\
\>  \\
\>   type Boolean = Either () () \\
\>  \\
\>   true, false :: Boolean \\
\>   true = Left () \\
\>   false = Right () \\
\>   \\
\>   ifthenelse :: Boolean -> a -> a  -> a \\
\>   ifthenelse b e1 e2 = either ($\backslash$() -> e1) ($\backslash$() -> e2) b 
\end{program*}
Here, the type {\tt{Boolean}} is defined as a disjoint union of the single
element unit type.  So, there are four elements, 

\begin{program*}
\> Boolean = \{Left(), Right(), $\bot$, Left $\bot$, Right $\bot$\} 
\end{program*}
The proper values of the type are {\tt{Left ()}} and {\tt{Right ()}}.

Now, we define constants {\tt{true}} and {\tt{false}} to be {\tt{Left ()}} and
{\tt{Right ()}} respectively.

The definition of {\tt{ifthenelse}} is somewhat interesting.  The function
{\tt{either}} is expecting functions as its first and seconf arguments. We want
{\tt{ifthenelse}} to just evaluate to the expression {\tt{e1}} if the condition
{\tt{b == true}} (recall {\tt{true}} is defined to be {\tt{Left()}}) and to
evaluate to {\tt{e2}} if the conditon {\tt{b==false}} (where {\tt{false}} has
been defined as {\tt{Right()}}).  We wrap {\tt{e1}} and {\tt{e2}} up as lambda
expressions whose argument is the constant {\tt{()}} {]em{i.e.}} the single
element of unit.  This may seem odd -- but the interpreter is using pattern
matching -- and since the only that that can be unwrapped by {\em{either}} if
the argument {\tt{b}} is {\tt{true}} or {\tt{false}} is {\tt{()}}, these
functions {\tt{($\backslash$() -> e1)}} and {\tt{($\backslash$() -> e2)}} will
only ever be applied to the constant {\tt{()}}.\footnote{As an experiment, you
might try typing in the following expressions to the interpreter:
{\tt{($\backslash$1 -> 2)1}} and \goodbreak {\tt{($\backslash$3 -> 2)1}} to see what
Haskell does.}


\exercise{Using the definition of {\tt{ifthenelse, true}} and {\tt{false}} given, prove that 
\begin{program*}
\> ifthenelse true e1 e2 = e1 \\
\> ifthenelse false e1 e2 = e2 
\end{program*}
}


\exercise{Using definitions of {\tt{ifthenelse, true}} and {\tt{false}} given
above, extend the code to include definitions of {\tt{and}}
({\tt{/$\backslash$}}), or ({\tt{$\backslash$/}}),  and implies
({\tt{.=>}}). 
\begin{program*}
\>   (/$\backslash$),($\backslash$/),(.=>) :: Boolean -> Boolean -> Boolean
\end{program*}
You may use the destructor {\tt{either}} directly if you like, or define the
functions in terms of {\tt{ifthenelse}}.  There is a link on the hw page to
{\tt{hw9\_expected.txt}} which contains an example test run of your code showing
what the results should look like.}


\end{document}
% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "% "
% compile-command: "latex hw9; dvips hw9 -o hw9.ps; pdflatex hw9"
% fill-column:79
% End:




