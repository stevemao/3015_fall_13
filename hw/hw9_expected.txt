                 Department of Computer Science
                     University of Wyoming

  To change your password, you need to use yppasswd, instead of passwd.

  If you have problems or questions, please e-mail Jim Ward (seker@uwyo.edu)
  or Allyson Anderson (allyson@uwyo.edu)

  For information on Printing, OpenOffice, JBuilder, Java, Firefox, 
      Eclipse, and other installed applications.  command: csinfo

  9/25/06: Eclipse has been updated to v3.2.  cmd: eclipse 
    If you want still want to use v3.1  cmd /usr/local/eclipse3.1/eclipse

done.
alpamayo 101% pwd
pwd
/home/faculty/jlc/www/courses/3015/hw
alpamayo 102% hugs
hugs
__   __ __  __  ____   ___      _________________________________________
||   || ||  || ||  || ||__      Hugs 98: Based on the Haskell 98 standard
||___|| ||__|| ||__||  __||     Copyright (c) 1994-2005
||---||         ___||           World Wide Web: http://haskell.org/hugs
||   ||                         Report bugs to: hugs-bugs@haskell.org
||   || Version: May 2006       _________________________________________

Haskell 98 mode: Restart with command line option -98 to enable extensions

Type :? for help
Hugs> :l hw9
:l hw9
Boolean> true /\ true
true /\ true
Left ()
Boolean> true /\ false
true /\ false
Right ()
Boolean> false /\ true
false /\ true
Right ()
Boolean> false /\ false
false /\ false
Right ()
Boolean> true \/ true
true \/ true
Left ()
Boolean> true \/ false
true \/ false
Left ()
Boolean> false \/ true
false \/ true
Left ()
Boolean> false \/ false
false \/ false
Right ()
Boolean> true .=> true
true .=> true
Left ()
Boolean> true .=> false
true .=> false
Right ()
Boolean> false  .=> true
false  .=> true
Left ()
Boolean> false .=> false
false .=> false
Left ()
Boolean> 